### PR TITLE
[nomnigraph] Move nomnigraph to OSS

### PR DESCRIPTION
**Pull request by bwasti**
This commit moves nomnigraph into caffe2/core and sets up basic cmake files for building the library.

### CODE REVIEW

File: caffe2/CMakeLists.txt [hunk 1]
```diff
@@ -115,7 +115,10 @@ add_library(caffe2 ${Caffe2_CPU_SRCS})
 target_link_libraries(caffe2 PUBLIC caffe2_protos ${Caffe2_PUBLIC_DEPENDENCY_LIBS})
 target_link_libraries(caffe2 PRIVATE ${Caffe2_DEPENDENCY_LIBS})
 target_link_libraries(caffe2 PRIVATE ${Caffe2_DEPENDENCY_WHOLE_LINK_LIBS})
-target_include_directories(caffe2 INTERFACE $<INSTALL_INTERFACE:include>)
+target_include_directories(caffe2
```
Comment (Maratyszcza) [inline caffe2/CMakeLists.txt]: This looks strange, what are you trying to do?

Comment (bwasti) [reply caffe2/CMakeLists.txt]: it is a way to get files inside `caffe2/core/nomnigraph/` to have `caffe2/core/nomnigraph/include` in their header namespace

Comment (bwasti) [reply caffe2/CMakeLists.txt]: I've simplified the build drastically by making nomnigraph a separate target

File: caffe2/CMakeLists.txt [hunk 2]
```diff
@@ -8,6 +8,7 @@ add_subdirectory(proto)
 
 add_subdirectory(contrib)
 add_subdirectory(core)
+add_subdirectory(core/nomnigraph)
```
Comment (Maratyszcza) [inline caffe2/CMakeLists.txt]: Is nomnigraph supposed to be an integral part of Caffe2? If not, consider adding `USE_NOMNIGRAPH` option in top-level CMakeLists

File: caffe2/core/nomnigraph/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,11 @@
+# ---[ CPU files.
+file(GLOB_RECURSE NOMNI_SRCS *.cc)
+file(GLOB NOMNI_TEST_SRCS test.cc)
+exclude(NOMNI_SRCS "${NOMNI_SRCS}" "${NOMNI_TEST_SRCS}")
+
+add_library(nomnigraph STATIC "${NOMNI_SRCS}")
```
Comment (Maratyszcza) [inline caffe2/core/nomnigraph/CMakeLists.txt]: `add_library(... OBJECT ...)` may be better if you don't expect any external projects to link against nomnigraph (and thus don't need `libnomnigraph.a`)

File: caffe2/core/nomnigraph/Converters/Caffe2.cc [hunk 4]
```diff
@@ -0,0 +1,327 @@
+#include "nomnigraph/Converters/Caffe2.h"
+#include "nomnigraph/Graph/Algorithms.h"
+
+#include "nomnigraph/Support/Casting.h"
+#include "nomnigraph/Support/Pointer.h"
+
+std::map<std::string, caffe2::Argument>
+getArgumentsFromOperator(caffe2::OperatorDef op) {
+  std::map<std::string, caffe2::Argument> argMap;
```
Comment (Maratyszcza) [inline caffe2/core/nomnigraph/Converters/Caffe2.cc]: I'm not sure building a map `arg name -> arg object` doesn't loose information: probably you could have multiple arguments with the same name in place of an argument with repeated value. I haven't seen such models, but have seen code that handles such situations. cc @Yangqing to clarify

Comment (Maratyszcza) [reply caffe2/core/nomnigraph/Converters/Caffe2.cc]: If doing an `arg name -> arg object` map  is fine, this function probably should be in `utils/proto_utils.h`

File: caffe2/core/nomnigraph/Converters/Caffe2.cc [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+    } else if (argMap.count("kernels")) {
+      for (auto i : argMap["kernels"].ints()) {
+        kernelShape.push_back(static_cast<int>(i));
+      }
+    } else if (argMap.count("kernel_h") && argMap.count("kernel_w")) {
+      int kernelH = static_cast<int>(argMap["kernel_h"].i());
+      int kernelW = static_cast<int>(argMap["kernel_w"].i());
+      kernelShape = {kernelH, kernelW};
+    } else {
+      assert(0);
```
Comment (Maratyszcza) [inline caffe2/core/nomnigraph/Converters/Caffe2.cc]: Please use `CAFFE_THROW`. `assert` is a no-op in Release mode, which is now default.

File: caffe2/core/nomnigraph/Converters/Caffe2.cc [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+      c->setStrides({stride, stride});
+    }
+
+    if (argMap.count("pad")) {
+      int pad = static_cast<int>(argMap["pad"].i());
+      c->setPads({pad, pad, pad, pad});
+    }
+
+    if (argMap.count("dialation")) {
+      int dialation = static_cast<int>(argMap["dialation"].i());
```
Comment (Maratyszcza) [inline caffe2/core/nomnigraph/Converters/Caffe2.cc]: typo!

File: caffe2/core/nomnigraph/Converters/Caffe2.cc [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+
+std::unique_ptr<repr::NeuralNetOperator>
+convertOperator(caffe2::OperatorDef op) {
+  auto argMap = getArgumentsFromOperator(op);
+
+  if (op.type() == "Conv") {
+    // There are literally three ways to define shapes in Conv in Caffe2
+    std::vector<int> kernelShape;
+    if (argMap.count("kernel")) {
+      int kernel = static_cast<int>(argMap["kernel"].i());
```
Comment (Maratyszcza) [inline caffe2/core/nomnigraph/Converters/Caffe2.cc]: It is worth to add e.g. `CAFFE_ENFORCE(argMap["kernel"].has_i(), "Error message");` before all lines that access `ArgumentProto` fields. Technically, these fields are optional in the message.

File: caffe2/core/nomnigraph/include/nomnigraph/Converters/Dot.h [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+  std::string convert(NodePrinter nodePrinter, EdgePrinter edgePrinter) {
+    std::ostringstream output;
+    output << "digraph G {\n\
+      bgcolor=\"#ffffff00\"\n\
+      color=white\n\
+      node[fontcolor=white,color=white];\n\
+      edge[fontcolor=white,color=white];\n\
+      ";
+    for (const auto &node : g_->getMutableNodes()) {
+      output << (uint64_t)node; // dot doesn't like hex
```
Comment (Maratyszcza) [inline caffe2/core/nomnigraph/include/nomnigraph/Converters/Dot.h]: Why C-style cast?

File: caffe2/core/nomnigraph/include/nomnigraph/Support/Casting.h [hunk 9]
```diff
@@ -0,0 +1,416 @@
...
+//
+// This file defines the isa<X>() function for checking downcastibility.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef NOM_SUPPORT_CASTING_H
+#define NOM_SUPPORT_CASTING_H
+
+#include <memory>
+#include <assert.h>
```
Comment (Maratyszcza) [inline caffe2/core/nomnigraph/include/nomnigraph/Support/Casting.h]: Please use `CAFFE_ENFORCE` instead

File: caffe2/core/nomnigraph/include/nomnigraph/Transformations/OperatorFusion.h [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+#define NOM_TRANSFORMATIONS_OPERATORFUSION_H
+
+#include "nomnigraph/Graph/Graph.h"
+#include "nomnigraph/Representations/NeuralNet.h"
+
+namespace nom {
+namespace transformations {
+
+bool fuseConvRelu(Graph<std::unique_ptr<repr::Value>, int> *);
+// std::unique_ptr<repr::NeuralNetData>> *);
```
Comment (Maratyszcza) [inline caffe2/core/nomnigraph/include/nomnigraph/Transformations/OperatorFusion.h]: dead code?

File: caffe2/core/nomnigraph/test.cc [hunk 11]
```diff
@@ -0,0 +1,692 @@
...
+#include "nomnigraph/Transformations/OperatorFusion.h"
+#include "nomnigraph/Transformations/Match.h"
+
+#include "nomnigraph/Support/Casting.h"
+
+#include <fstream>
+#include <iomanip>
+#include <stdio.h>
+
+#define ADD_ARG(_op, _name, _type, _val)                                       \
```
Comment (Maratyszcza) [inline caffe2/core/nomnigraph/test.cc]: You could get away with just two overloaded functions to add s(tring) and i(nt) arguments

File: caffe2/core/nomnigraph/test.cc [hunk 12]
```diff
@@ -0,0 +1,692 @@
...
+    labelMap["shape"] = "box";
+  } else if (isa<nom::repr::Data>(node->data())) {
+    auto tensor = dyn_cast<nom::repr::NeuralNetData>(node->data().get());
+    labelMap["label"] = tensor->getName();
+    labelMap["label"] += "_" + std::to_string(tensor->getVersion()) + " " + std::to_string((unsigned long long)node);
+  }
+  return labelMap;
+};
+
+int main(int argc, char *argv[]) {
```
Comment (Maratyszcza) [inline caffe2/core/nomnigraph/test.cc]: Please use Google Test for the unit tests

**Review by Maratyszcza — APPROVED** [review]
LGTM, but please fix comments before merging

File: caffe2/CMakeLists.txt (hunk 2 above)
Comment (bwasti) [reply caffe2/CMakeLists.txt]: it is supposed to be an integral part of caffe2

File: caffe2/core/nomnigraph/Converters/Caffe2.cc (hunk 4 above)
Comment (bwasti) [reply caffe2/core/nomnigraph/Converters/Caffe2.cc]: this is how Glow does it I believe

File: caffe2/core/nomnigraph/Converters/Caffe2.cc (hunk 6 above)
Comment (bwasti) [reply caffe2/core/nomnigraph/Converters/Caffe2.cc]: 😅

File: caffe2/core/nomnigraph/test.cc (hunk 12 above)
Comment (bwasti) [reply caffe2/core/nomnigraph/test.cc]: this whole test file is going to be formalized in a later PR (and actually migrated to caffe2 test locations in caffe2/opt).  I'll remove it for now

File: caffe2/core/nomnigraph/Converters/Caffe2.cc (hunk 5 above)
Comment (bwasti) [reply caffe2/core/nomnigraph/Converters/Caffe2.cc]: this choice was inspired by the LLVM codebase, which doesn't use exceptions to reduce codesize and potentially speed up the program

**Pull request closed without merging**